core/src/replay_stage.rs

```diff
@@ -404,6 +422,221 @@ impl ReplayStage {
+            the duplicate slots to choose from",
+                );
+            Some(bank)
+        } else {
+            None
+        }
+    }
+
+    // Checks for and handle forks with duplicate slots.
+    fn check_for_duplicate_slots(
+        my_pubkey: &Pubkey,
+        duplicate_slots_receiver: &DuplicateSlotReceiver,
+        ancestors: &HashMap<Slot, HashSet<Slot>>,
+        progress: &mut ProgressMap,
+        blockstore: &Blockstore,
+        bank_forks: &RwLock<BankForks>,
+        poh_recorder: &Mutex<PohRecorder>,
+        tower: &Tower,
+        leader_schedule_cache: &LeaderScheduleCache,
+    ) {
+        let mut unconfirmed_forks = vec![];
+        for duplicate_slot in duplicate_slots_receiver.try_iter() {
+            if blockstore.is_root(duplicate_slot) {
+                Self::handle_approved_duplicate_blockhash(
+                    blockstore,
+                    duplicate_slot,
+                    &Hash::default(),
+                );
+            } else {
+                if let Some(fork_progress) = progress.get(&duplicate_slot) {
+                    if fork_progress.fork_stats.confirmation_reported {
+                        Self::handle_approved_duplicate_blockhash(
+                            blockstore,
+                            duplicate_slot,
+                            &Hash::default(),
+                        );
+                    } else {
+                        Self::add_unconfirmed_duplicate_slot(
+                            duplicate_slot,
+                            &mut unconfirmed_forks,
+                            ancestors,
+                        );
+                    }
+                };
+
+                // Mark the duplicate as dead, signaling RepairService to query
+                // the cluster about alternate versions of this slot
+                blockstore
+                    .set_dead_slot(duplicate_slot)
+                    .expect("Failed to mark slot as dead in blockstore");
+
+                // Figure out if the validator needs to reset PoH
+                let last_reset_slot = poh_recorder.lock().unwrap().last_reset_slot();
+                if let Some(reset_bank) = Self::process_duplicate_unconfirmed_forks(
+                    my_pubkey,
+                    ancestors,
+                    &unconfirmed_forks,
+                    progress,
+                    bank_forks,
+                    last_reset_slot,
+                    tower,
+                ) {
+                    Self::reset_poh_recorder(
+                        &my_pubkey,
+                        &blockstore,
+                        &reset_bank,
+                        poh_recorder,
+                        &leader_schedule_cache,
+                    )
+                };
+            }
+        }
+    }
+
+    fn handle_approved_duplicate_blockhash(
+        _blockstore: &Blockstore,
+        _slot: Slot,
+        _blockhash: &Hash,
+    ) {
+    }
```

Comment:
@sakridge This will get filled in with the RepairService changes once we've updated the DeadSlots column to include an approved_blockhash